components/kyma-environment-broker/internal/process/provisioning/inject_btp_operator_credentials_step.go

```diff
@@ -0,0 +1,145 @@
+package provisioning
+
+import (
+	"context"
+	"reflect"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal"
+	kebError "github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/error"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/process"
+	"github.com/kyma-project/control-plane/components/kyma-environment-broker/internal/storage"
+	"github.com/sirupsen/logrus"
+	apicorev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	updateSecretBackoff = 10 * time.Second
+	secretName          = "sap-btp-manager"
+	secretNamespace     = "kyma-system"
+)
+
+var labels = map[string]string{"app.kubernetes.io/managed-by": "kcp-kyma-environment-broker"}
+var annotations = map[string]string{"Warning": "This secret is generated. Do not edit!"}
+
+type InjectBTPOperatorCredentialsStep struct {
+	operationManager  *process.OperationManager
+	k8sClientProvider func(kubeconfig string) (client.Client, error)
+}
+
+func NewInjectBTPOperatorCredentialsStep(os storage.Operations, k8sClientProvider func(kcfg string) (client.Client, error)) *InjectBTPOperatorCredentialsStep {
+	return &InjectBTPOperatorCredentialsStep{
+		operationManager:  process.NewOperationManager(os),
+		k8sClientProvider: k8sClientProvider,
+	}
+}
+
+func (s *InjectBTPOperatorCredentialsStep) Name() string {
+	return "Inject_BTP_Operator_Credentials"
+}
+
+func (s *InjectBTPOperatorCredentialsStep) Run(operation internal.Operation, log logrus.FieldLogger) (internal.Operation, time.Duration, error) {
+
+	if operation.RuntimeID == "" {
+		log.Error("Runtime ID is empty")
+		return s.operationManager.OperationFailed(operation, "Runtime ID is empty", nil, log)
+	}
+
+	clusterID := operation.InstanceDetails.ServiceManagerClusterID
+	if clusterID == "" {
+		clusterID = uuid.NewString()
+		updatedOperation, backoff, _ := s.operationManager.UpdateOperation(operation, func(op *internal.Operation) {
+			op.InstanceDetails.ServiceManagerClusterID = clusterID
+		}, log)
+		if backoff != 0 {
+			log.Error("cannot save cluster ID")
+			return updatedOperation, backoff, nil
+		}
+	}
+
+	if operation.ProvisioningParameters.ErsContext.SMOperatorCredentials == nil {
+		log.Error("no credentials set")
+		return s.operationManager.OperationFailed(operation, "no credentials set", nil, log)
+	}
+
+	secret := s.prepareSecret(operation.ProvisioningParameters.ErsContext.SMOperatorCredentials, clusterID)
+
+	if operation.K8sClient == nil {
```

Comment:
nit: really small performance optimization would be to move all these guards for operation prerequisites to the top of `Run()`, because at the moment, KEB would potentially do some database queries related to `ServiceManagerClusterID` and then later find out k8s client set is missing and fail the whole operation.

Reply:
Moved in the next commit.